connectivity/suite.go

```diff
@@ -250,6 +250,89 @@ func Run(ctx context.Context, ct *check.ConnectivityTest, addExtraTests func(*ch
 	}
 
+	if ct.Params().Short {
+		addShortSuite(ct, renderedTemplates)
+	} else {
+		addFullSuite(ct, renderedTemplates)
+	}
+
+	// Tests with DNS redirects to the proxy (e.g., client-egress-l7, dns-only,
+	// and to-fqdns) should always be executed last. See #367 for details.
+
+	if err := addExtraTests(ct); err != nil {
+		return err
+	}
+
+	if ct.Params().IncludeUnsafeTests {
+		ct.NewTest("check-log-errors").WithScenarios(tests.NoErrorsInLogs(ct.CiliumVersion))
+	}
+
+	return ct.Run(ctx)
+}
+
+// The main idea of --short is to give a quick signal whether basic features of
+// Cilium work as expected. --short is used by the upgrade GH workflows in
+// order to reduce their runtime.
+//
+// List of things to test (optimized for low runtime):
+// * pod-to-pod, pod-to-host, pod-to-service
+// * no-netpol / L3+L4-netpol / L7-netpol
+// * N/S LB
+func addShortSuite(ct *check.ConnectivityTest, renderedTemplates map[string]string) {
+	// Run all tests without any policies in place.
+	noPoliciesScenarios := []check.Scenario{
+		tests.PodToPod(),
+		tests.ClientToClient(),
+		tests.PodToService(),
+		tests.PodToWorld(tests.WithRetryAll()),
+		tests.PodToHost(),
+	}
+
+	ct.NewTest("no-policies").WithScenarios(noPoliciesScenarios...)
+
+	ct.NewTest("echo-ingress").WithCiliumPolicy(echoIngressFromOtherClientPolicyYAML).
+		WithScenarios(tests.PodToPod()).
+		WithExpectations(func(a *check.Action) (egress, ingress check.Result) {
+			if a.Destination().HasLabel("kind", "echo") && !a.Source().HasLabel("other", "client") {
+				// TCP handshake fails both in egress and ingress when
+				// L3(/L4) policy drops at either location.
+				return check.ResultDropCurlTimeout, check.ResultDropCurlTimeout
+			}
+			return check.ResultOK, check.ResultOK
+		})
+
+	// Test L7 HTTP introspection using an ingress policy on echo pods.
+	ct.NewTest("echo-ingress-l7").
+		WithFeatureRequirements(features.RequireEnabled(features.L7Proxy)).
+		WithCiliumPolicy(echoIngressL7HTTPPolicyYAML). // L7 allow policy with HTTP introspection
+		WithScenarios(
+			tests.PodToPodWithEndpoints(),
+		).
+		WithExpectations(func(a *check.Action) (egress, ingress check.Result) {
+			if a.Source().HasLabel("other", "client") { // Only client2 is allowed to make HTTP calls.
+				// Trying to access private endpoint without "secret" header set
+				// should lead to a drop.
+				if a.Destination().Path() == "/private" && !a.Destination().HasLabel("X-Very-Secret-Token", "42") {
+					return check.ResultDropCurlHTTPError, check.ResultNone
+				}
+				egress = check.ResultOK
+				// Expect all curls from client2 to be proxied and to be GET calls.
+				egress.HTTP = check.HTTP{
+					Method: "GET",
+				}
+				return egress, check.ResultNone
+			}
+			return check.ResultDrop, check.ResultDefaultDenyIngressDrop
+		})
+
+	ct.NewTest("north-south-loadbalancing").
+		WithFeatureRequirements(features.RequireEnabled(features.NodeWithoutCilium)).
+		WithScenarios(
+			tests.OutsideToNodePort(),
+		)
```

Comment:
Do you think there would be a way to avoid duplicating these test definitions, i.e. make the long tests a superset of the short tests?

Reply:
Good question. I'm not sure how to structure the tests that they could be reused. Any ideas?

I thought about annotating each Test/Scenarion with `.Short()`, but this idea felt short when I saw how many changes are needed to implement this :-/

Reply:
> Good question. I'm not sure how to structure the tests that they could be reused. Any ideas?
> 
> I thought about annotating each Test/Scenarion with `.Short()`, but this idea felt short when I saw how many changes are needed to implement this :-/

Yeah, something like this would be my preference. I had a quick go at it in a slightly different way (completely untested though):

```diff
diff --git a/connectivity/check/context.go b/connectivity/check/context.go
index 0353f51ad74f..c9bb6b314788 100644
--- a/connectivity/check/context.go
+++ b/connectivity/check/context.go
@@ -243,6 +243,14 @@ func NewConnectivityTest(client *k8s.Client, p Parameters, version string) (*Con
 // a new Test. This object can be used to set up the environment to execute
 // different Scenarios within.
 func (ct *ConnectivityTest) NewTest(name string) *Test {
+       return ct.newTest(name, false)
+}
+
+func (ct *ConnectivityTest) NewShortTest(name string) *Test {
+       return ct.newTest(name, true)
+}
+
+func (ct *ConnectivityTest) newTest(name string, short bool) *Test {
        var member struct{}
 
        if name == "" {
@@ -263,6 +271,7 @@ func (ct *ConnectivityTest) NewTest(name string) *Test {
                verbose:   ct.verbose(),
                logBuf:    &bytes.Buffer{}, // maintain internal buffer by default
                warnBuf:   &bytes.Buffer{},
+               short:     short,
        }
 
        // Setting the internal buffer to nil causes the logger to
diff --git a/connectivity/check/test.go b/connectivity/check/test.go
index b9a21bbad0ce..f8b1aa3855e0 100644
--- a/connectivity/check/test.go
+++ b/connectivity/check/test.go
@@ -67,6 +67,9 @@ type Test struct {
        // True if the Test is marked as failed.
        failed bool
 
+       // True if the Test should be run as part of the short test suite.
+       short bool
+
        // requirements is a list of required Cilium features which need to match
        // for this test to be run
        requirements []features.Requirement
@@ -232,6 +235,12 @@ func (t *Test) willRun() (bool, string) {
                return false, reason
        }
 
+       // Skip the Test if it is not marked as a short test and we're running
+       // only the short test suite.
+       if !t.short && t.Context().params.Short {
+               return false, "skipped in short test suite"
+       }
+
        // Skip the whole Test if all of its Scenarios are excluded by the user's
        // filter.
        var skipped int
```

and then declare all the short tests using `NewShortTest` instead of `NewTest` (alternatively we could still implement `.Short()`. Would something like this work?

Reply:
Thinking about it a bit more, adding `.Short()` instead of `NewShortTest` is probably nicer.

```diff
diff --git a/connectivity/check/test.go b/connectivity/check/test.go
index b9a21bbad0ce..eda37aac9001 100644
--- a/connectivity/check/test.go
+++ b/connectivity/check/test.go
@@ -67,6 +67,9 @@ type Test struct {
        // True if the Test is marked as failed.
        failed bool
 
+       // True if the Test should be run as part of the short test suite.
+       short bool
+
        // requirements is a list of required Cilium features which need to match
        // for this test to be run
        requirements []features.Requirement
@@ -232,6 +235,12 @@ func (t *Test) willRun() (bool, string) {
                return false, reason
        }
 
+       // Skip the Test if it is not marked as a short test and we're running
+       // only the short test suite.
+       if !t.short && t.Context().params.Short {
+               return false, "skipped in short test suite"
+       }
+
        // Skip the whole Test if all of its Scenarios are excluded by the user's
        // filter.
        var skipped int
@@ -336,6 +345,12 @@ func (t *Test) Run(ctx context.Context) error {
        return nil
 }
 
+// Short marks the test to be run in the short test suite only.
+func (t *Test) Short() *Test {
+       t.short = true
+       return t
+}
+
 // WithCiliumPolicy takes a string containing a YAML policy document and adds
 // the polic(y)(ies) to the scope of the Test, to be applied when the test
 // starts running. When calling this method, note that the CNP enabled feature
```

Tested locally with marking the `no-policies` test with `.Short()`:

```
🏃 Running tests...
[=] Test [no-policies]
..............................
[=] Skipping Test [no-policies-extra] (skipped in short test suite)
[=] Skipping Test [allow-all-except-world] (skipped in short test suite)
[=] Skipping Test [client-ingress] (skipped in short test suite)
[=] Skipping Test [client-ingress-knp] (skipped in short test suite)
[=] Skipping Test [allow-all-with-metrics-check] (skipped in short test suite)
[=] Skipping Test [all-ingress-deny] (skipped in short test suite)
[=] Skipping Test [all-ingress-deny-knp] (skipped in short test suite)
[=] Skipping Test [all-egress-deny] (skipped in short test suite)
[=] Skipping Test [all-egress-deny-knp] (skipped in short test suite)
[=] Skipping Test [all-entities-deny] (skipped in short test suite)
[=] Skipping Test [cluster-entity] (skipped in short test suite)
[=] Skipping Test [host-entity] (skipped in short test suite)
[=] Skipping Test [echo-ingress] (skipped in short test suite)
[=] Skipping Test [echo-ingress-knp] (skipped in short test suite)
[=] Skipping Test [client-ingress-icmp] (skipped in short test suite)
[=] Skipping Test [client-egress] (skipped in short test suite)
[=] Skipping Test [client-egress-knp] (skipped in short test suite)
[=] Skipping Test [client-egress-expression] (skipped in short test suite)
[=] Skipping Test [client-egress-expression-knp] (skipped in short test suite)
[=] Skipping Test [client-with-service-account-egress-to-echo] (skipped in short test suite)
[=] Skipping Test [client-egress-to-echo-service-account] (skipped in short test suite)
[=] Skipping Test [to-entities-world] (skipped in short test suite)
[=] Skipping Test [to-cidr-external] (skipped in short test suite)
[=] Skipping Test [to-cidr-external-knp] (skipped in short test suite)
[=] Skipping Test [echo-ingress-from-other-client-deny] (skipped in short test suite)
[=] Skipping Test [client-ingress-from-other-client-icmp-deny] (skipped in short test suite)
[=] Skipping Test [client-egress-to-echo-deny] (skipped in short test suite)
[=] Skipping Test [client-ingress-to-echo-named-port-deny] (skipped in short test suite)
[=] Skipping Test [client-egress-to-echo-expression-deny] (skipped in short test suite)
[=] Skipping Test [client-with-service-account-egress-to-echo-deny] (skipped in short test suite)
[=] Skipping Test [client-egress-to-echo-service-account-deny] (skipped in short test suite)
[=] Skipping Test [client-egress-to-cidr-deny] (skipped in short test suite)
[=] Skipping Test [client-egress-to-cidr-deny-default] (skipped in short test suite)
[=] Skipping Test [health] (skipped in short test suite)
[=] Skipping Test [north-south-loadbalancing] (Feature node-without-cilium is disabled)
[=] Skipping Test [pod-to-pod-encryption] (skipped in short test suite)
[=] Skipping Test [node-to-node-encryption] (skipped in short test suite)
[=] Skipping Test [egress-gateway-excluded-cidrs] (Feature enable-ipv4-egress-gateway is disabled)
[=] Skipping Test [pod-to-node-cidrpolicy] (Feature cidr-match-nodes is disabled)
[=] Skipping Test [north-south-loadbalancing-with-l7-policy] (Feature node-without-cilium is disabled)
[=] Skipping Test [echo-ingress-l7] (skipped in short test suite)
[=] Skipping Test [echo-ingress-l7-named-port] (skipped in short test suite)
[=] Skipping Test [client-egress-l7-method] (skipped in short test suite)
[=] Skipping Test [client-egress-l7] (skipped in short test suite)
[=] Skipping Test [client-egress-l7-named-port] (skipped in short test suite)
[=] Skipping Test [client-egress-l7-tls-deny-without-headers] (Feature secret-backend-k8s is disabled)
[=] Skipping Test [client-egress-l7-tls-headers] (Feature secret-backend-k8s is disabled)
[=] Skipping Test [client-egress-l7-set-header] (Feature secret-backend-k8s is disabled)
[=] Skipping Test [echo-ingress-auth-always-fail] (Feature mutual-auth-spiffe is disabled)
[=] Skipping Test [echo-ingress-mutual-auth-spiffe] (Feature mutual-auth-spiffe is disabled)
[=] Skipping Test [pod-to-ingress-service] (Feature ingress-controller is disabled)
[=] Skipping Test [pod-to-ingress-service-deny-all] (Feature ingress-controller is disabled)
[=] Skipping Test [pod-to-ingress-service-allow-ingress-identity] (Feature ingress-controller is disabled)
[=] Skipping Test [dns-only] (skipped in short test suite)
[=] Skipping Test [to-fqdns] (skipped in short test suite)

✅ All 1 tests (30 actions) successful, 55 tests skipped, 0 scenarios skipped.
```

Reply:
Thanks! That's what I had originally. But then started adding the same annotation method to `Scenario` and quickly gave up. I think we can live for now with the annotation only for `Test`.

Do you mind if I keep `addFullSuite()` commit for better readability of the `Run` function?

Reply:
Yeah, keeping addFullSuite makes sense for readability 👍